### PR TITLE
[dev.v1] backport v2 machinery

### PR DESCRIPTION
**Pull request by leoluk**
This backports most v2 features to v1:

- Signed heartbeats with network state.
- Chain error counters.
- Guardian set state with a list-nodes CLI command.
- Observation requests for Ethereum.
- New governance injection format with multiple messages per file.
- Zap logging setup improvements.

Includes a few small build system backports:

- Use stable Buildkit Dockerfile syntax.
- Use `buf` config for proto generation.
- Updated tools to dev.v2 versions.

Upgraded Go dependencies to the same versions as dev.v2 and replaced
d-fuse by gagliardetto Solana RPC client.

Removes a bunch of dead Terra code.

The v1 Solana build is dead due to broken external dependencies (yay),
so this was MacGyver-ed to use the mainnet binary artifact instead.

Tried splitting this up into multiple PRs, but it's a lot of
interdependent code that's hard to take apart such that each
commit still compiles.

E2E tests succeeded (with four nodes):

```
--- PASS: TestEndToEnd_SOL_ETH (46.13s)
--- PASS: TestEndToEnd_SOL_ETH/[SOL]_Native_->_[ETH]_Wrapped (19.97s)
--- PASS: TestEndToEnd_SOL_ETH/[ETH]_Wrapped_->_[SOL]_Native (3.49s)
--- PASS: TestEndToEnd_SOL_ETH/[ETH]_Native_->_[SOL]_Wrapped (3.52s)
--- PASS: TestEndToEnd_SOL_ETH/[SOL]_Wrapped_->_[ETH]_Native (19.13s)
```

list-nodes:

```
$ kc exec -it guardian-1 -- /guardiand admin list-nodes --socket /tmp/admin.sock
Defaulted container "guardiand" out of: guardiand, agent
2022/02/16 21:11:12 current guardian set index: 1 (5 guardians)
2022/02/16 21:11:12 5 nodes in guardian state set
Node key                                              Guardian key                                Node name   Version      Last seen     Solana  Ethereum
12D3KooWL3XJ9EMCyZvmmGXL2LMiVBtrVa2BuESsJiXkSj7333Jw  0xbeFA429d57cD18b7F8A4d91A2da9AB4AF05d0FBe  guardian-0  development  9.623515311s  376     171
12D3KooWHHzSeKaY8xuZVzkLbKFfvNgPPeKhFBGrMbNzbm5akpqu  0x88D7D8B32a9105d228100E72dFFe2Fae0705D31c  guardian-1  development  1.849686114s  396     178
12D3KooWKRyzVWW6ChFjQjK4miCty85Niy49tpPV95XdKu1BcvMA  0x58076F561CC62A47087B567C86f986426dFCD000  guardian-2  development  1.829417513s  396     178
12D3KooWB1b3qZxWJanuhtseF3DmPggHCtG36KZ9ixkqHtdKH9fh  0xBd6e9833490F8fA87c733A183CD076a6cBD29074  guardian-3  development  1.857038787s  396     178
12D3KooWE4qDcRrueTuRYWUdQZgcy7APZqBngVeXRt4Y6ytHizKV  0xb853FCF0a5C78C1b56D15fCE7a154e6ebe9ED7a2  guardian-4  development  9.208957972s  378     171
```

Manually tested Eth re-observation requests and verified that
they result in the same digest and successfully re-observe the lockup.

Please review carefully in spite of the large number of changes!

---
This change is [Reviewable](https://reviewable.io/reviews/certusone/wormhole/869)

### CODE REVIEW

File: node/pkg/vaa/structs.go [hunk 1]
```diff
@@ -94,34 +95,53 @@ type (
 	}
 )
 
+func (a Address) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(`"%s"`, a)), nil
```
Comment (SEJeff) [inline node/pkg/vaa/structs.go:99]: Seems fine but if it is a string, why not use `%q`?

File: solana/Dockerfile [hunk 2]
```diff
@@ -1,27 +1,32 @@
...
+# COPY Cargo.lock .
+#
+# RUN --mount=type=cache,target=/usr/local,from=rust,source=/usr/local \
+#     --mount=type=cache,target=/root/.cache \
+#     --mount=type=cache,target=bridge/target \
+#     --mount=type=cache,target=target \
+#     --mount=type=cache,target=bin,from=rust,source=bin \
+#     cargo build-bpf --manifest-path "bridge/Cargo.toml" && \
+#     mkdir -p /opt/solana/deps && \
+#     cp target/deploy/spl_bridge.so /opt/solana/deps/spl_bridge.so
```
Comment (SEJeff) [inline solana/Dockerfile:32]: Dead code, remove it?

File: Tiltfile [hunk 3]
```diff
@@ -59,10 +59,15 @@ def k8s_yaml_with_ns(objects):
...
 
+proto_deps = ["./proto", "buf.yaml", "buf.gen.yaml"]
+
 local_resource(
     name = "proto-gen",
-    deps = ["./proto", "./generate-protos.sh"],
-    cmd = "./generate-protos.sh",
+    deps = proto_deps,
+    cmd = "tilt docker build -- --target go-export -f Dockerfile.proto -o type=local,dest=node .",
+    env = {"DOCKER_BUILDKIT": "1"},
```
[the remark this answers is not in the document]
Comment (evan-gray) [reply Tiltfile:68]: Output from dockerfiles = :heart:
https://github.com/moby/buildkit#output

File: Dockerfile.proto [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# syntax=docker.io/docker/dockerfile:1.3@sha256:42399d4635eddd7a9b8a24be879d2f9a930d0ed040a61324cfdf59ef1357b3b2
+FROM docker.io/golang:1.17.5@sha256:90d1ab81f3d157ca649a9ff8d251691b810d95ea6023a03cdca139df58bca599 AS go-tools
+
+RUN mkdir /app
+
+ADD tools/build.sh /app/tools/
+ADD tools/go.* /app/tools/
```
[the remark this answers is not in the document]
Comment (leoluk) [reply Dockerfile.proto:7]: Wait that's valid syntax? Thought you could only copy one thing at the time.

[the remark this answers is not in the document]
Comment (leoluk) [reply Dockerfile.proto:7]: Wait that's valid syntax? :O

File: node/cmd/guardiand/adminnodes.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+			if showDetails {
+				fields = append(fields, fmt.Sprintf("%s %d (%d)",
+					truncAddrs[n.ChainID], heights[n.ChainID], errors[n.ChainID]))
+			} else {
+				fields = append(fields, fmt.Sprintf("%d", heights[n.ChainID]))
+			}
+		}
+
+		for _, field := range fields {
+			_, _ = fmt.Fprintf(w, "%s\t", field)
```
[the remark this answers is not in the document]
Comment (leoluk) [reply node/cmd/guardiand/adminnodes.go:156]: Silences linter warnings

File: Dockerfile.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (SEJeff) [reply Dockerfile.proto:7]:
```
$ cat Dockerfile
FROM docker.io/centos:8
COPY 1 2 /data/

$ podman build -t wtf .
STEP 1/2: FROM docker.io/centos:8
STEP 2/2: COPY 1 2 /data/
COMMIT wtf
--> 461b746496a
Successfully tagged localhost/wtf:latest
461b746496ab6a36c317ebd027db83cc46d6b1bcf26452705897e31492b55397

$ podman run -it wtf ls /data/
1  2
```

File: solana/Dockerfile (hunk 2 above)
Comment (leoluk) [reply solana/Dockerfile:32]: It's not really dead, the build is just broken and needs someone to fix it. Not supposed to stay this way.

**Review by evan-gray — APPROVED** [review]
> Reviewed 107 of 107 files at r1.
> *[Reviewable](https://reviewable.io/reviews/certusone/wormhole/869)* status: 106 of 107 files reviewed, 26 unresolved discussions (waiting on @evan-gray, @hendrikhofstadt, @leoluk, and @SEJeff)

___
*[node/cmd/guardiand/bridge.go, line 176 at r1](https://reviewable.io/reviews/certusone/wormhole/869#-Mw7fI-F7UDf_jm258dJ:-Mw7fI-F7UDf_jm258dK:b-kpwn4s) ([raw file](https://github.com/certusone/wormhole/blob/4c492a5132851cbb51d264b52469a2dacf6ec9f0/node/cmd/guardiand/bridge.go#L176)):*
> ```Go
> 
> 	// Redirect ipfs logs to plain zap
> 	ipfslog.SetPrimaryCore(logger.Core())
> ```

Will this logger change modify the logging output in a way that may break anyones monitoring?
___
*[proto/gossip/v1/gossip.proto, line 5 at r1](https://reviewable.io/reviews/certusone/wormhole/869#-Mw8-ytL3uK3szBdNMyO:-Mw8-ytL3uK3szBdNMyP:b-fuy2o2) ([raw file](https://github.com/certusone/wormhole/blob/4c492a5132851cbb51d264b52469a2dacf6ec9f0/proto/gossip/v1/gossip.proto#L5)):*
> ```protobuf
> package gossip.v1;
> 
> option go_package = "github.com/certusone/wormhole/node/pkg/proto/gossip/v1;gossipv1";
> ```

does it matter that this change is only intended for a branch of that github link? just changing for more proper namespacing?

**Review by evan-gray — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/certusone/wormhole/869)* status: all files reviewed, 26 unresolved discussions (waiting on @hendrikhofstadt, @leoluk, and @SEJeff)

**Review by SEJeff — APPROVED** [review]
nits to be addressed in a followup, but there are no hard blockers here. Looks great 👍

**Review by SEJeff — APPROVED** [review]

**Pull request merged**